### PR TITLE
First package split in dvote-js

### PR DESCRIPTION
**Pull request by marcvelmer**
Two new packages created:
 - `common`: Should be finished and ready for npm publish
 - `encryption`: Several files missing like CHANGELOG, README, AUTHORS, LICENSE

The `encryption` package has a dependency from `common` via `common.tgz` file which should be changed for real npm published package.

~~Still not working: Tests in `dvote-js` (Error when trying to find `vocdoni-common` package)~~ Already fixed by referencing packages locally. Adding also TODO for easy replacement when packages are published.

### CODE REVIEW

File: packages/common/README.md [hunk 1]
```diff
@@ -0,0 +1,41 @@
+# @vocdoni/common
+
+@vocdoni/common is the common package of the [dvote-js library](https://github.com/vocdoni/dvote-js/)
```
Comment (brickpop) [inline packages/common/README.md]:
```suggestion
@vocdoni/common contains shared helpers, models and type definitions for the [dvote-js library](https://github.com/vocdoni/dvote-js/)
```

File: packages/common/test/mocha-hooks.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+let testSuites = 0
```
Comment (brickpop) [inline packages/common/test/mocha-hooks.ts]: This file should be moved to a common folder (`/common/test/mocha-hooks.ts`), and not be copied on each subpackage.

File: packages/common/tslint.json [hunk 3]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (brickpop) [inline packages/common/tslint.json]: This file may be redundant (from a code editor's perspective). Only the root one might be enough?

File: src/api/census.ts [hunk 4]
```diff
@@ -1,7 +1,7 @@
 import { Wallet, Signer, providers, BigNumber, ContractReceipt, Contract } from "ethers"
 import { IRequestParameters } from "../net/gateway-dvote"
 import { Keccak256, Poseidon } from "../crypto/hashing"
-import { hexStringToBuffer } from "../util/encoding"
+import { hexStringToBuffer } from "../../packages/common/src"
```
Comment (brickpop) [inline src/api/census.ts]:
```suggestion
import { hexStringToBuffer } from "../../packages/common/src"  // TODO: Import from the new NPM package
```

File: src/api/voting.ts [hunk 5]
```diff
@@ -12,7 +12,7 @@ import {
 } from "../constants"
 import { BytesSignature } from "../crypto/data-signing"
 import { Buffer } from "buffer/"  // Previously using "arraybuffer-to-string"
-import { Asymmetric } from "../crypto/encryption"
+import { Asymmetric } from "../../packages/encryption/src"
```
Comment (brickpop) [inline src/api/voting.ts]:
```suggestion
import { Asymmetric } from "../../packages/encryption/src"  // TODO: Import from the new NPM package
```

File: src/api/voting.ts [hunk 6]
```diff
@@ -36,7 +36,7 @@ import { CensusErc20Api } from "./census"
 import { ProcessEnvelopeType } from "dvote-solidity"
 import { IGatewayClient, IGatewayDVoteClient, IGatewayWeb3Client } from "../common"
 import { Poseidon } from "../crypto/hashing"
-import { uintArrayToHex } from "../util/encoding"
+import { uintArrayToHex } from "../../packages/common/src"
```
Comment (brickpop) [inline src/api/voting.ts]:
```suggestion
import { uintArrayToHex } from "../../packages/common/src"  // TODO: Import from the new NPM package
```

File: src/crypto/wallets.ts [hunk 7]
```diff
@@ -5,13 +5,13 @@ import { eddsa, babyJub } from "circomlib"
 // import { Scalar, utils as ffutils } from "ffjavascript"
 // const { Scalar, utils: ffutils } = require("../../node_modules/ffjavascript/build/main.cjs")
 const { Scalar, utils: ffutils } = require("ffjavascript")
-import { bufferToBigInt } from "../util/encoding"
+import { bufferToBigInt } from "../../packages/common/src"
```
Comment (brickpop) [inline src/crypto/wallets.ts]:
```suggestion
import { bufferToBigInt } from "../../packages/common/src"  // TODO: Import from the new NPM package
```

File: src/index.ts [hunk 8]
```diff
@@ -28,7 +28,7 @@ export * from "./net/ipfs"
 // CRYPTO
 export * from "./crypto/data-signing"
 export * from "./crypto/elliptic"
-export * from "./crypto/encryption"
+export * from "../packages/encryption/src"
```
Comment (brickpop) [inline src/index.ts]:
```suggestion
export * from "../packages/encryption/src"  // TODO: Export from the new NPM package
```

File: src/models/backup.ts [hunk 9]
```diff
@@ -1,4 +1,4 @@
-import { Symmetric } from "../crypto/encryption"
+import { Symmetric } from "../../packages/encryption/src"
```
Comment (brickpop) [inline src/models/backup.ts]:
```suggestion
import { Symmetric } from "../../packages/encryption/src"  // TODO: Import from the new NPM package
```

File: test/integration/processes.ts [hunk 10]
```diff
@@ -25,7 +25,7 @@ import NamespaceBuilder from "../builders/namespace"
 import { Web3Gateway } from "../../src/net/gateway-web3"
 import GenesisBuilder, { DEFAULT_CHAIN_ID } from "../builders/genesis"
 import ResultsBuilder from "../builders/results"
-import { VotingApi } from "../../dist"
+import { VotingApi } from "../../dist/src"
```
Comment (brickpop) [inline test/integration/processes.ts]: Why do we need this change? Isn't it exported from the dvote-js root module?

File: test/unit/census.ts [hunk 11]
```diff
@@ -5,7 +5,7 @@ import { Wallet, utils } from "ethers"
 import { Buffer } from "buffer/"
 
 import { CensusOffChain } from "../../src/api/census"
-import { compressPublicKey } from "../../dist"
+import { compressPublicKey } from "../../dist/src"
```
Comment (brickpop) [inline test/unit/census.ts]: Same as above

File: test/unit/data-signing.ts [hunk 12]
```diff
@@ -6,7 +6,7 @@ import { TextEncoder, TextDecoder } from "util"
 
 import { JsonSignature, BytesSignature } from "../../src/crypto/data-signing"
 import { extractUint8ArrayJSONValue } from "../../src/util/uint8array"
-import { compressPublicKey } from "../../dist"
+import { compressPublicKey } from "../../dist/src"
```
Comment (brickpop) [inline test/unit/data-signing.ts]: Same as above

File: test/unit/processes.ts [hunk 13]
```diff
@@ -13,11 +13,11 @@ import { GenesisContractMethods, ProcessCensusOrigin, ProcessesContractMethods,
 import { Buffer } from "buffer/"
 
 import { VotingApi, VotePackage } from "../../src/api/voting"
-import { Asymmetric } from "../../src/crypto/encryption"
+import { Asymmetric } from "../../packages/encryption/src"
```
Comment (brickpop) [inline test/unit/processes.ts]:
```suggestion
import { Asymmetric } from "../../packages/encryption/src"  // TODO: Import from the new NPM package
```

File: test/unit/wallets.ts [hunk 14]
```diff
@@ -6,7 +6,7 @@ import { utils } from "ethers"
 import { WalletBabyJub, WalletUtil } from "../../src/crypto/wallets"
 import { Random } from "../../src/util/random"
 import { JsonSignature, BytesSignature } from "../../src/crypto/data-signing"
-import { compressPublicKey } from "../../dist"
+import { compressPublicKey } from "../../dist/src"
```
Comment (brickpop) [inline test/unit/wallets.ts]: Same as above

File: packages/client/src/file.ts [hunk 15]
```diff
@@ -1,12 +1,14 @@
...
+    IGatewayDVoteClient,
+    IPFS,
+    IRequestParameters
+} from "../../net/src" // TODO reference the future package
 import { Buffer } from 'buffer/'
 import axios from "axios"
-import { Wallet, Signer } from "ethers"
-import { promiseWithTimeout } from "../util/timeout"
+import { Signer, Wallet } from "ethers"
+import { promiseWithTimeout } from "../../common/src" // TODO reference the future package
```
Comment (brickpop) [inline packages/client/src/file.ts]: promiseWithTimeout looks more like an utility than a common function

Comment (marcvelmer) [reply packages/client/src/file.ts]: You are right, but it's actually used in `net` and `client` package, therefore is in common. But totally agree that this should be considered as a util function. But I also have some doubts concerning this point that I want to discuss in a conversation :)

**Comment by brickpop** [comment]
Pending tasks: 
- [x] Adapt the E2E scripts on the `examples` folder
- [x] Edit the main readme
- [ ] Publish dvote-js@1.14.0
- [x] Update the GitHub actions script, to run the tests on each individual package

**Pull request merged**